swig/include/python/ogr_python.i

```diff
@@ -168,6 +168,100 @@ def _WarnIfUserHasNotSpecifiedIfUsingExceptions():
             raise TypeError("Input %s is not of String or Int type" % type(value))
 %}
 
+%feature("shadow") ExecuteSQL %{
+def ExecuteSQL(self, statement, spatialFilter=None, dialect="", keep_ref_on_ds=False):
+    """ExecuteSQL(self, statement, spatialFilter: ogr.Geometry = None, dialect: Optional[str] = "", keep_ref_on_ds=False) -> ogr.Layer
+
+    Execute a SQL statement against the dataset
+
+    The result of a SQL query is:
+      - None (or an exception if exceptions are enabled) for statements
+        that are in error
+      - or None for statements that have no results set,
+      - or a ogr.Layer handle representing a results set from the query.
+
+    Note that this ogr.Layer is in addition to the layers in the data store
+    and must be destroyed with ReleaseResultSet() before the data source is closed
+    (destroyed).
+
+    Starting with GDAL 3.7, this method can also be used as a context manager,
+    as a convenient way of automatically de-allocating the returned result layer.
+
+    For more information on the SQL dialect supported internally by OGR
+    review the OGR SQL document (:ref:`ogr_sql_sqlite_dialect`)
+    Some drivers (i.e. Oracle and PostGIS) pass the SQL directly through to the
+    underlying RDBMS.
+
+    The SQLITE dialect can also be used (:ref:`sql_sqlite_dialect`)
+
+    Parameters
+    ----------
+    statement:
+        the SQL statement to execute (e.g "SELECT * FROM layer")
+    spatialFilter:
+        a geometry which represents a spatial filter. Can be None
+    dialect:
+        allows control of the statement dialect. If set to None or empty string,
+        the OGR SQL engine will be used, except for RDBMS drivers that will
+        use their dedicated SQL engine, unless OGRSQL is explicitly passed as
+        the dialect. The SQLITE dialect can also be used.
+    keep_ref_on_ds:
+        whether the returned layer should keep a (strong) reference on
+        the current dataset. Cf example 2 for a use case.
+
+    Returns
+    -------
+    ogr.Layer:
+        a ogr.Layer containing the results of the query, that will be
+        automatically freed when the context manager goes out of scope
```

Comment:
We should be consistent with the verb used to indicate deallocation: "free" or "destroy" or one might think that they are referring to different operations.

Reply:
done. I've opted for released/releasing to align with the ReleaseResultSet() method